### PR TITLE
C2CX Passthrough using "market" orders

### PR DESCRIPTION
**Pull request by gz-c**
* Still needs manual integration tests
* Still needs a notes document describing unknown cases and failure possibilities and how to deal with them

### CODE REVIEW

File: src/exchange/passthrough.go [hunk 1]
```diff
@@ -186,170 +359,337 @@ func (p *Passthrough) handleDepositInfoState(di DepositInfo) (DepositInfo, error
...
-		// TODO -- use UpdateDepositInfoCallback, in case the DB save fails?
-		di, err = p.store.UpdateDepositInfo(di.DepositID, func(di DepositInfo) DepositInfo {
-			di.Passthrough.Orders = append(di.Passthrough.Orders, PassthroughOrder{})
-			// di.Passthrough.SkyBought += o.Amount
-			// di.Passthrough.DepositValueSpent += o.Amount * o.Price
-			return di
-		})
-		if err != nil {
-			return di, err
+			if order.CustomerID == nil || *order.CustomerID != di.Passthrough.Order.CustomerID {
```
Comment (therealssj) [inline src/exchange/passthrough.go:596]: split this into `CustomerID` missing and `CustomerID` not equal?

File: src/exchange/passthrough.go [hunk 2]
```diff
@@ -186,170 +359,337 @@ func (p *Passthrough) handleDepositInfoState(di DepositInfo) (DepositInfo, error
...
+func calculateRequestedAmount(depositValue int64) decimal.Decimal {
+	amount := decimal.New(depositValue, -int32(SatoshiExponent))
+	amount = amount.Truncate(int32(c2cx.TradePairRulesTable[c2cx.BtcSky].PricePrecision))
+	return amount
 }
 
-/* END PSEUDOCODE FOR EXCHANGE PURCHASE LOGIC */
+// calculateSkyBought returns the amount of SKY bought in droplets
+// The amount of SKY bought is in order.CompletedAmount
+// This amount does is not adjusted for the C2CX commission, which is not
```
Comment (therealssj) [inline src/exchange/passthrough.go]: typo

File: src/exchange/passthrough.go [hunk 3]
```diff
@@ -186,170 +359,337 @@ func (p *Passthrough) handleDepositInfoState(di DepositInfo) (DepositInfo, error
...
+// known through the API, so the actual amount bought is less.
+// For now, ignore the commission and eat the fee.
+func calculateSkyBought(order *c2cx.Order) (uint64, error) {
+	// Convert CompletedAmount from whole skycoin to satoshis
```
Comment (therealssj) [inline src/exchange/passthrough.go]: to droplets

File: src/exchange/passthrough_test.go [hunk 4]
```diff
@@ -0,0 +1,1134 @@
...
+		name   string
+		bought string
+		price  string
+		out    int64
+	}{
+		{
+			name:   "negative",
+			bought: "-1",
+			price:  "0.002",
+			out:    -200000,
```
Comment (therealssj) [inline src/exchange/passthrough_test.go:1085]: Why do we allow it to be negative?

**Review by therealssj — COMMENTED** [review]
looks solid

File: src/exchange/passthrough.go (hunk 1 above)
Comment (gz-c) [reply src/exchange/passthrough.go:596]: It should never happen and the full order data is in the log

File: src/exchange/passthrough_test.go (hunk 4 above)
Comment (gz-c) [reply src/exchange/passthrough_test.go:1085]: It should never be negative, just checking that the calculation doesn't freak out if it happened to see a negative value though

**Pull request merged**